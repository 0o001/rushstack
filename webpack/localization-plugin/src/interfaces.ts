// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * @public
 */
export interface IPassthroughLocaleOptions {
  usePassthroughLocale?: boolean;

  /**
   * If {@link IPassthroughLocaleOptions.usePassthroughLocale} is set, use this name for the passthrough locale.
   * Defaults to "passthrough"
   */
  passthroughLocaleName?: string;
}

/**
 * @public
 */
export interface ITypingsGenerationOptions {
  generatedTsFolder: string;
  sourceRoot?: string;
  exportAsDefault?: boolean;
}

/**
 * @public
 */
export interface IDefaultLocaleOptions {
  localeName: string;

  fillMissingTranslationStrings?: boolean;
}

/**
 * Options for the pseudolocale library.
 *
 * @internalRemarks
 * Eventually this should be replaced with DefinitelyTyped types.
 *
 * @public
 */
export interface IPseudolocaleOptions {
  prepend?: string;
  append?: string;
  delimiter?: string;
  startDelimiter?: string;
  endDelimiter?: string;
  extend?: number;
  override?: string;
}

/**
 * @public
 */
export interface IPseudolocalesOptions {
  [pseudoLocaleName: string]: IPseudolocaleOptions;
}

/**
 * @public
 */
export interface ILocalizedData {
  /**
   * Use this parameter to specify the translated data.
   */
  translatedStrings: ILocalizedStrings;

  /**
   * Options around including a passthrough locale.
   */
  passthroughLocale?: IPassthroughLocaleOptions;

  /**
   * Options for the locale used in the source localized data files.
   */
  defaultLocale: IDefaultLocaleOptions;

  /**
   * Options for pseudo-localization.
   */
  pseudolocales?: IPseudolocalesOptions;
}

/**
 * @public
 */
export interface ILocalizationStatsOptions {
  dropPath?: string;

  callback?: (stats: ILocalizationStats) => void;
}

/**
 * The options for localization.
 *
 * @public
 */
export interface ILocalizationPluginOptions {
  localizedData: ILocalizedData;

  filesToIgnore?: string[];

  localizationStats?: ILocalizationStatsOptions;

  typingsOptions?: ITypingsGenerationOptions;
}

/**
 * @internal
 */
export interface ILocalizationFile {
  [stringName: string]: ILocalizedString;
}

/**
 * @internal
 */
export interface ILocalizedString {
  value: string;
  comment?: string;
}

/**
 * @public
 */
export interface ILocaleFileData {
  [stringName: string]: string;
}

/**
 * @public
 */
export interface ILocaleData {
  [locFilePath: string]: ILocaleFileData;
}

/**
 * @public
 */
export interface ILocalizedStrings {
  [locale: string]: ILocaleData;
}


/**
 * @public
 */
export interface ILocalizationStatsEntrypoint {
  localizedAssets: { [locale: string]: string };
}

/**
 * @public
 */
export interface ILocalizationStatsChunkGroup {
  localizedAssets: { [locale: string]: string };
}

/**
 * @public
 */
export interface ILocalizationStats {
  entrypoints: { [name: string]: ILocalizationStatsEntrypoint };
  namedChunkGroups: { [name: string]: ILocalizationStatsChunkGroup };
}
