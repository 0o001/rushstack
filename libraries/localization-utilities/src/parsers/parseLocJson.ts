// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { resolve } from 'path';

import { JsonFile, JsonSchema } from '@rushstack/node-core-library';

import { ILocalizationFile, IParseFileOptions } from '../interfaces';

const LOC_JSON_SCHEMA_PATH: string = resolve(__dirname, '../schemas/locJson.schema.json');
const LOC_JSON_SCHEMA: JsonSchema = JsonSchema.fromFile(LOC_JSON_SCHEMA_PATH);

/**
 * @public
 */
export function parseLocJson({ content, filePath, ignoreString }: IParseFileOptions): ILocalizationFile {
  const parsedFile: ILocalizationFile = JsonFile.parseString(content);
  try {
    LOC_JSON_SCHEMA.validateObject(parsedFile, filePath);
  } catch (e) {
    throw new Error(`The loc file is invalid. Error: ${e}`);
  }

  if (ignoreString) {
    const newParsedFile: ILocalizationFile = {};
    for (const [key, stringData] of Object.entries(parsedFile)) {
      if (!ignoreString(filePath, key)) {
        newParsedFile[key] = stringData;
      }
    }

    return newParsedFile;
  } else {
    return parsedFile;
  }
}
