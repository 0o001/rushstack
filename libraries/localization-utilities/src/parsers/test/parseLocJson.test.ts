// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { IgnoreStringFunction } from '../../interfaces';
import { parseLocJson } from '../parseLocJson';
import { serializeLocFile } from './serializeLocFile';

describe(parseLocJson.name, () => {
  it('parses a valid file', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        comment: 'A string'
      },
      bar: {
        value: 'Bar',
        comment: 'Another string'
      }
    });

    expect(
      serializeLocFile(
        parseLocJson({
          content,
          filePath: 'test.loc.json'
        })
      )
    ).toMatchInlineSnapshot(`
      "key | value | comment
      ------------------------------
       foo | Foo   | A string
       bar | Bar   | Another string"
    `);
  });

  it('throws on invalid file', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        baz: 'A string'
      }
    });

    expect(() =>
      parseLocJson({
        content,
        filePath: 'test.loc.json'
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it('correctly ignores a string', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        comment: 'A string'
      },
      bar: {
        value: 'Bar',
        comment: 'Another string'
      }
    });

    const ignoredStringFunction: IgnoreStringFunction = jest
      .fn()
      .mockImplementation(
        (fileName: string, stringName: string) => fileName === 'test.loc.json' && stringName === 'bar'
      );

    expect(
      serializeLocFile(
        parseLocJson({
          content,
          filePath: 'test.loc.json',
          ignoreString: ignoredStringFunction as IgnoreStringFunction
        })
      )
    ).toMatchInlineSnapshot(`
      "key | value | comment
      ------------------------
       foo | Foo   | A string"
    `);

    expect((ignoredStringFunction as unknown as jest.SpyInstance).mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "test.loc.json",
          "foo",
        ],
        Array [
          "test.loc.json",
          "bar",
        ],
      ]
    `);
  });
});
