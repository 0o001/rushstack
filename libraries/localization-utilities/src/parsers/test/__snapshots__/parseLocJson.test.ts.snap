// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseLocJson throws on invalid file 1`] = `
"The loc file is invalid. Error: Error: JSON validation failed:
test.loc.json

Error: #/foo
       Additional properties not allowed: baz"
`;
